src/Neo/UInt160.cs

```diff
@@ -31,68 +32,72 @@ public class UInt160 : IComparable<UInt160>, IEquatable<UInt160>, ISerializable
+        [FieldOffset(16)] private uint _value3;
 
         public int Size => Length;
 
         /// <summary>
         /// Initializes a new instance of the <see cref="UInt160"/> class.
         /// </summary>
-        public UInt160()
-        {
-        }
+        public UInt160() { }
 
         /// <summary>
         /// Initializes a new instance of the <see cref="UInt160"/> class.
         /// </summary>
         /// <param name="value">The value of the <see cref="UInt160"/>.</param>
-        public unsafe UInt160(ReadOnlySpan<byte> value)
+        public UInt160(ReadOnlySpan<byte> value)
         {
-            if (value.Length != Length) throw new FormatException();
-            fixed (ulong* p = &value1)
-            {
-                Span<byte> dst = new(p, Length);
-                value[..Length].CopyTo(dst);
-            }
+            if (value.Length != Length)
+                throw new FormatException();
+
+            var bytes = value.ToArray();
+            _value1 = Unsafe.As<byte, ulong>(ref bytes[0]);
+            _value2 = Unsafe.As<byte, ulong>(ref bytes[8]);
+            _value3 = Unsafe.As<byte, uint>(ref bytes[16]);
         }
 
         public int CompareTo(UInt160 other)
         {
-            int result = value3.CompareTo(other.value3);
+            var result = _value3.CompareTo(other._value3);
             if (result != 0) return result;
-            result = value2.CompareTo(other.value2);
+            result = _value2.CompareTo(other._value2);
             if (result != 0) return result;
-            return value1.CompareTo(other.value1);
+            return _value1.CompareTo(other._value1);
         }
 
         public void Deserialize(ref MemoryReader reader)
         {
-            value1 = reader.ReadUInt64();
-            value2 = reader.ReadUInt64();
-            value3 = reader.ReadUInt32();
+            _value1 = reader.ReadUInt64();
+            _value2 = reader.ReadUInt64();
+            _value3 = reader.ReadUInt32();
         }
 
         public override bool Equals(object obj)
         {
-            if (ReferenceEquals(obj, this)) return true;
-            return Equals(obj as UInt160);
+            if (obj == null)
+                return false;
+
+            var other = obj as UInt160;
+            if (other == null)
+                return false;
+            return Equals(other);
         }
 
         public bool Equals(UInt160 other)
         {
-            if (other is null) return false;
-            return value1 == other.value1
-                && value2 == other.value2
-                && value3 == other.value3;
+            if (other == null) return false;
+            return _value1 == other._value1 &&
+                _value2 == other._value2 &&
+                _value3 == other._value3;
         }
 
         public override int GetHashCode()
         {
-            return (int)value1;
+            return HashCode.Combine(_value1, _value2, _value3);
```

Comment:
I am not sure if this has any side affect since this is a public method.  @shargon  how do you think?

Reply:
That's not how `GetHashCode` works. The `HashCode` has to be the same each time. When the data is the same. Just think about it, what's the name of the method? Hash... duh, data must hash the same without collisions. Old way was incorrect and a bug.

Reply:
there will be collision, and highly possible. hsahcode is only 4 bytes.

Reply:
If there is a collision, the equal method is called, nothing importart I think.

Reply:
![image](https://github.com/user-attachments/assets/9068e091-0d2a-4ed7-82be-62bcdda17f50)

source: https://learn.microsoft.com/en-us/dotnet/fundamentals/runtime-libraries/system-object-gethashcode